[IR] Add elementwise modifier to atomicrmw (#189517)
llvm/llvm-project: llvm/docs/LangRef.rst

For xchg, this
may also be a floating point or a pointer type with the same size constraints
as integers.  For fadd/fsub/fmax/fmin/fmaximum/fminimum/fmaximumnum/fminimumnum, this must be a floating-point

> or fixed vector of floating-point type, with the same size constraints. The type of the '``<pointer>``'

Reviewer: Why was the prior wording here wrong?
What does "same size constraint" mean here? (should it be "with the same size constraints as integers"?)

operand must be a pointer to that type. If the ``atomicrmw`` is marked
as ``volatile``, then the optimizer is not allowed to modify the
number or order of execution of this ``atomicrmw`` with other
:ref:`volatile operations <volatile>`. If the ``elementwise`` modifier is present,

> then ``<value>`` must be a fixed vector type, with the same size constraints, whose element type is legal for the

Reviewer: "same size constraints as integers"?

An ``atomicrmw`` instruction can also take an optional
":ref:`syncscope <syncscope>`" argument.

If the ``elementwise`` modifier is present, the instruction has per-element vector

> atomic semantics. It behaves as if it were expanded into one scalar ``atomicrmw`` per element, executed in an arbitrary order.

Reviewer: Maybe:

```suggestion
atomic semantics. It behaves as if it were expanded into one scalar ``atomicrmw`` per element that are not ordered with respect to each other.
```

? 

@jyknight thoughts?

I think the core of what we want to say is that there are no ordering guarantees among the scalar per-element `atomicrmw`. This matters a lot when a memory ordering different than `unordered` or `monotonic` is used.

Reviewer 2: Sorry, I missed this discussion when it originally happened! I do agree that _unordered_ between the elements, rather than simply arbitrary order, is a preferable specification.

But, I'm afraid this will require more changes to fix up other parts of the specification after adding this new "internally unordered" concept.

E.g., if I have an `atomicrmw elementwise ptr %p, <2 x i8> %val seq_cst`, then, 2 other reader threads could potentially observe the stores happening in opposite orders...which seq_cst's total ordering typically prohibits. I think that is what we do mean to say, BUT, it's not well supported by the LangRef descriptions.

ISTM that the impact is very similar to that of partially-overlapping or mixed-size atomic operations, which we purport to support but haven't specified, and is already under discussion on #204329, so I think we can attempt to resolve the specification issues introduced here as part of that discussion.

Without ``elementwise``, vector ``atomicrmw`` keeps whole-value atomic semantics.

Targets may implement ``atomicrmw elementwise`` either by lowering it to a
native elementwise vector atomic, by scalarizing it into per-element scalar
``atomicrmw`` operations, or by using an existing stronger whole-value atomic
implementation, as long as the observable semantics are at least as strong as

> the IR definition above.

Reviewer: Is this paragraph needed? There are other implementation strategies available, and that all those strategies are valid as long as they conform to the IR specification above is something that is universally true for the entire LangRef.
